src/InputFileResolver.ts

```diff
@@ -4,96 +4,143 @@ import * as _ from 'lodash';
 import * as log4js from 'log4js';
 
 const log = log4js.getLogger('InputFileResolver');
+
 const DEFAULT_INPUT_FILE_PROPERTIES = { mutated: false, included: true };
 
+
 export default class InputFileResolver {
 
-  private inputFileDescriptors: InputFileDescriptor[];
-  private mutateFileExpressions: string[];
+  private inputFileResolver: PatternResolver;
+  private mutateResolver: PatternResolver;
 
   constructor(mutate: string[], allFileExpressions: Array<InputFileDescriptor | string>) {
-    this.mutateFileExpressions = mutate || [];
-    this.inputFileDescriptors = allFileExpressions.map(maybePattern => {
-      if (InputFileResolver.isInputFileDescriptor(maybePattern)) {
-        return maybePattern;
-      } else {
-        return <InputFileDescriptor>_.assign({ pattern: maybePattern }, DEFAULT_INPUT_FILE_PROPERTIES);
-      }
-    });
+    validateFileDescriptor(allFileExpressions);
+    this.mutateResolver = PatternResolver.parse(mutate || []);
+    this.inputFileResolver = PatternResolver.parse(allFileExpressions);
   }
 
   public resolve(): Promise<InputFile[]> {
-    let mutateFilePromise = this.resolveMutateFileGlobs();
-    return this.resolveInputFileGlobs().then((allInputFiles) => mutateFilePromise.then(additionalMutateFiles => {
-      InputFileResolver.markAdditionalFilesToMutate(allInputFiles, additionalMutateFiles);
-      InputFileResolver.warnAboutNoFilesToMutate(allInputFiles);
-      return allInputFiles;
-    }));
+    return Promise.all([this.inputFileResolver.resolve(), this.mutateResolver.resolve()]).then(results => {
+      const inputFiles = results[0];
+      const mutateFiles = results[1];
+      markAdditionalFilesToMutate(inputFiles, mutateFiles.map(m => m.path));
+      logFilesToMutate(inputFiles);
+      return inputFiles;
+    });
   }
+}
+
+class PatternResolver {
+
+  private ignore = false;
+  private descriptor: InputFileDescriptor;
 
-  private static markAdditionalFilesToMutate(allInputFiles: InputFile[], additionalMutateFiles: string[]) {
-    let errors: string[] = [];
-    additionalMutateFiles.forEach(mutateFile => {
-      if (!allInputFiles.filter(inputFile => inputFile.path === mutateFile).length) {
-        errors.push(`Could not find mutate file "${mutateFile}" in list of files.`);
+  constructor(descriptor: InputFileDescriptor | string, private previous?: PatternResolver) {
```

Comment:
Why did you choose an implementation that gave every PatternResolver a link to the previous one? For me, it would make more sense when the InputFileResolver would combine each of the PatternResolvers, while the PatternResolver now does it himself.